### PR TITLE
feat: add image transformations, cropping, and color adjustments

### PR DESCRIPTION
**Pull request by PatrykC01**
## Summary
- Adds image transformations in the edit window: rotate 90° left/right, rotate 180°, horizontal flip, and vertical flip.
- Adds freeform cropping and improves DPI ruler calibration/scaling.
- Adds pixel-based image adjustments: color inversion, Photoshop/GIMP-like brightness and contrast behavior, and desaturation with color-range sliders.
- Updates edit preview/save flow so effects are rendered through canvas and saved as real pixel changes instead of CSS-only filters.

## Test plan
- Verified TypeScript check passes during pre-commit.
- Verified unit tests pass during pre-commit.
- Manually tested brightness/contrast edge cases against GIMP behavior.
- Manually tested cropping, transformations, DPI scaling, inversion, and desaturation in the edit window.

### CODE REVIEW

File: src/components/edit-window/edit-window.tsx [hunk 1]
```diff
@@ -323,15 +421,202 @@ export function EditWindow() {
...
+                    ctx.imageSmoothingQuality = "low";
+                    ctx.drawImage(source, 0, 0, canvas.width, canvas.height);
+                    return replaceBaseImageFromCanvas(canvas).then(() => {
+                        const scaleText = scaleFactor.toFixed(3);
+                        if (
+                            canvas.width === sourceWidth &&
+                            canvas.height === sourceHeight
+                        ) {
+                            toast.info(
+                                `DPI: skala ${scaleText}x, ` +
```
Comment (TheMultii) [inline src/components/edit-window/edit-window.tsx]: hardcoded pl. use i18next

File: src/components/edit-window/edit-window.tsx [hunk 2]
```diff
@@ -323,15 +421,202 @@ export function EditWindow() {
...
+                                    `rozmiar bez zmian (${canvas.width} x ` +
+                                    `${canvas.height} px)`
+                            );
+                            return;
+                        }
+
+                        toast.success(
+                            `DPI: skala ${scaleText}x, ` +
```
Comment (TheMultii) [inline src/components/edit-window/edit-window.tsx]: hardcoded pl. use i18next

File: src/components/edit-window/edit-window.tsx [hunk 3]
```diff
@@ -323,15 +421,202 @@ export function EditWindow() {
...
+                    ctx.drawImage(source, 0, 0, canvas.width, canvas.height);
+                    return replaceBaseImageFromCanvas(canvas).then(() => {
+                        const scaleText = scaleFactor.toFixed(3);
+                        if (
+                            canvas.width === sourceWidth &&
+                            canvas.height === sourceHeight
+                        ) {
+                            toast.info(
+                                `DPI: skala ${scaleText}x, ` +
+                                    `rozmiar bez zmian (${canvas.width} x ` +
```
Comment (TheMultii) [inline src/components/edit-window/edit-window.tsx]: hardcoded pl. use i18next

File: src/lib/imageModifiers/pipeline.ts [hunk 4]
```diff
@@ -50,37 +184,51 @@ export async function applyPipelineToImage(
...
+        .find(
+            (mod): mod is BrightnessModifier =>
+                mod.enabled && mod.type === "brightness"
+        );
+    const contrastModifier = [...modifiers]
+        .reverse()
+        .find(
+            (mod): mod is ContrastModifier =>
+                mod.enabled && mod.type === "contrast"
+        );
```
Comment (TheMultii) [inline src/lib/imageModifiers/pipeline.ts]: Duplicate modifiers are silently dropped

File: src/components/edit-window/modifiers/ModifierSettingsDialog.tsx [hunk 5]
```diff
@@ -399,7 +531,7 @@ function FftSettings({
         if (!maskCvs) return;
         const ctx = maskCvs.getContext("2d");
         ctx?.clearRect(0, 0, maskCvs.width, maskCvs.height);
-        onChange({ _maskCanvas: maskCvs });
+        onChange({ _maskCanvas: maskCvs, maskDataUrl: null });
```
Comment (TheMultii) [inline src/components/edit-window/modifiers/ModifierSettingsDialog.tsx]: am i wrong or is maskDataUrl always null?

File: src/components/edit-window/edit-window.tsx [hunk 6]
```diff
@@ -5,23 +5,35 @@ import { Menubar } from "@/components/ui/menubar";
...
     buildCssFilter,
+    hasCanvasModifiers,
 } from "@/lib/imageModifiers/registry";
 import { applyPipelineToImage } from "@/lib/imageModifiers/pipeline";
 import { AddModifierButton } from "@/components/edit-window/modifiers/AddModifierButton";
 import { ModifierList } from "@/components/edit-window/modifiers/ModifierList";
 import { ModifierSettingsDialog } from "@/components/edit-window/modifiers/ModifierSettingsDialog";
 
+const CANVAS_CONTEXT_UNAVAILABLE = "Canvas context unavailable";
+const FAILED_TO_SAVE_IMAGE_KEY = "Failed to save image: {{error}}";
```
Comment (TheMultii) [inline src/components/edit-window/edit-window.tsx:35]: A failed rotation/crop is not a save failure - use operation-specific keys

File: src/components/edit-window/dpi/image-dpi-controls.tsx [hunk 7]
```diff
@@ -96,6 +107,29 @@ export default function ImageDpiControls({
                     ))}
                 </div>
             </div>
+
+            <div className="space-y-1">
+                <span className="block text-sm font-medium">
+                    Reference length (mm)
```
Comment (TheMultii) [inline src/components/edit-window/dpi/image-dpi-controls.tsx]: hardcoded

File: src/components/edit-window/edit-window.tsx [hunk 8]
```diff
@@ -278,7 +376,7 @@ export function EditWindow() {
         const newMod = def.create() as AnyModifier;
         setModifiers(prev => [...prev, newMod]);
         // Automatically open edit dialog for the new modifier
-        setEditingModifierId(newMod.id);
+        setTimeout(() => setEditingModifierId(newMod.id), 50);
```
Comment (TheMultii) [inline src/components/edit-window/edit-window.tsx]: This looks like a race workaround

**Review by TheMultii — CHANGES_REQUESTED** [review]

**Comment by TheMultii** [comment]
@PatrykC01 resolve conflicts
